File: Data Structure and Algorithms/assignment/LinkedList/list.lua
function CreateNode(value)
    return {
        value = value,
        next = nil,
    }
end

local LinkedList = {}
LinkedList.__index = LinkedList

function LinkedList.new()
    return setmetatable({head = nil}, LinkedList);
end

function LinkedList:print()
    local current = self.head;
    while current do
        io.write(tostring(current.value));
        current = current.next;

        if current ~= nil then
            io.write("->")
        end
    end

    io.write("\n");
end

function LinkedList:add(value)
    if self.head == nil then
        self.head = CreateNode(value);
    else
        local current = self.head;
        while current ~= nil and current.next ~= nil do
            current = current.next;
        end

        current.next = CreateNode(value);
    end
end

function LinkedList:remove(value)
    if not self.head then return end;

    if self.head.value == value then
        self.head = self.head.next;
        return;
    end

    local current = self.head;

    while current ~= nil and current.next ~= nil and current.next.value ~= value do
        current = current.next;
    end

    if current and current.next then
        current.next = current.next.next;
        return;
    end

end

function LinkedList:sort(func)
    if self.head == nil then return end;

    repeat
        local swapped = false;
        local current = self.head;

        while current ~= nil and current.next ~= nil do
            if func(current.value, current.next.value) then
                current.value, current.next.value = current.next.value, current.value;
                swapped = true;
            end

            current = current.next;
        end

    until not swapped
end

function LinkedList:insert_start(value)
    local node = CreateNode(value);
    node.next = self.head;
    self.head = node;
end

function LinkedList:insert_after(value, pos)
    if pos < 1 then return end;

    local newNode = CreateNode(value);
    local temp = self.head;

    if pos == 0 then
        self:insert_start(value);
        return;
    end

    for _ = 1, pos - 1 do
        if not temp then return end;
        temp = temp.next;
    end

    if not temp then return end;
    newNode.next = temp.next;
    temp.next = newNode;
end

function LinkedList:insert_before(value, pos)
    local new_node = CreateNode(value);
    local current_pos = 1;
    local current = self.head;
    local prev;

    while current and current_pos < pos do
        prev = current;
        current = current.next;
        current_pos = current_pos + 1;
    end

    if not current then return end;
    prev.next = new_node;
    new_node.next = current;
end

function LinkedList:getLength()
    if not self.head then return 0; end
    local current = self.head;
    local length = 1;

    while current ~= nil and current.next ~= nil do
        length = length + 1;
        current = current.next;
    end

    return length;
end

function LinkedList:remove_duplicates()
    if not self.head then return end;

    local current = self.head;

    while current do
        local runner = current;

        while runner and runner.next do
            if runner.next.value == current.value then
                runner.next = runner.next.next
            else
                runner = runner.next;
            end
        end

        current = current.next;
    end

end

function LinkedList:clear()
    self.head = nil;
end

return LinkedList;

File: Data Structure and Algorithms/assignment/LinkedList/list.js
function Node(data) {
    return {
        data: data,
        next: null,
    }
}

class LinkedList {
    constructor() {
        this.head = null;
    }

    insert_beginning(data) {
        const new_node = Node(data);
        new_node.next = this.head;
        this.head = new_node;
    }

    insert_end(data) {
        const new_node = Node(data);

        if (!this.head) {
            this.head = new_node;
        } else {
            let temp = this.head;

            while (temp.next) {
                temp = temp.next;
            }

            temp.next = new_node;
        }
    }

    insert_after_position(data, position) {
        if (position < 0) return;

        if (position === 0) {
            this.insert_beginning(data);
            return;
        }

        const new_node = Node(data);
        let temp = this.head;
        let count = 1;

        while (temp && count < position) {
            temp = temp.next;
            count++;
        }

        if (!temp) return;

        new_node.next = temp.next
        temp.next = new_node;
    }

    insert_before_position(data, position) {
        if (position < 0) return;

        if (position === 0) {
            this.insert_beginning(data);
            return;
        }

        const new_node = Node(data);
        let temp = this.head;
        let count = 1;

        while (temp && count < position - 1) {
            temp = temp.next;
            count++;
        }

        if (!temp) return;
        new_node.next = temp.next;
        temp.next = new_node;
    }


    print() {
        let temp = this.head;
        let str = "";

        while (temp) {
            str += temp.data + " ";
            temp = temp.next;
        }

       console.log(`[${str.trim()}]`);
    }
}

const list = new LinkedList();

console.log("starting")
list.insert_beginning(2)
list.insert_end(1)
list.insert_end(3)
list.insert_end(4)
list.insert_after_position(5, 1)
list.insert_before_position(6, 3)
list.insert_before_position(7, 4)

list.print()